contracts/utils/cryptography/draft-ERC7739Utils.sol

```diff
@@ -0,0 +1,288 @@
+    function unwrapTypedDataSig(
+        bytes calldata wrappedSignature
+    )
+        internal
+        pure
+        returns (bytes calldata signature, bytes32 separator, bytes32 contents, bytes calldata contentsType)
+    {
+        unchecked {
+            uint256 sigLength = wrappedSignature.length;
+
+            if (sigLength < 66) return (_emptyCalldataBytes(), 0, 0, _emptyCalldataBytes());
+
+            uint256 contentsTypeEnd = sigLength - 2; // Last 2 bytes
+            uint256 contentsTypeLength = uint16(bytes2(wrappedSignature[contentsTypeEnd:]));
+
+            if (contentsTypeLength > contentsTypeEnd) return (_emptyCalldataBytes(), 0, 0, _emptyCalldataBytes());
+
+            uint256 contentsEnd = contentsTypeEnd - contentsTypeLength;
+
+            if (contentsEnd < 64) return (_emptyCalldataBytes(), 0, 0, _emptyCalldataBytes());
+
+            uint256 separatorEnd = contentsEnd - 32;
+            uint256 signatureEnd = separatorEnd - 32;
+
+            signature = wrappedSignature[:signatureEnd];
+            separator = bytes32(wrappedSignature[signatureEnd:separatorEnd]);
+            contents = bytes32(wrappedSignature[separatorEnd:contentsEnd]);
+            contentsType = wrappedSignature[contentsEnd:contentsTypeEnd];
+        }
+    }
+
+    /**
+     * @dev Nest a signature for a given EIP-712 type into a nested signature for the domain `separator`.
+     *
+     * Counterpart of {unwrapTypedDataSig} to extract the original signature and the nested components.
+     */
+    function wrapTypedDataSig(
+        bytes memory signature,
+        bytes32 separator,
+        bytes32 contents,
+        bytes memory contentsType
+    ) internal pure returns (bytes memory) {
+        return abi.encodePacked(signature, separator, contents, contentsType, uint16(contentsType.length));
+    }
+
+    /**
+     * @dev Nests a `contents` digest into an {EIP712} type that simulates the `eth_personalSign` RPC
+     * method in the context of smart contracts.
+     *
+     * This typed uses the {_NESTED_PERSONAL_SIGN_TYPEHASH} type to nest the `contents` hash for the
+     * current domain `separator`.
+     *
+     * To produce a signature for this nested type, the signer must sign the following type hash:
+     *
+     * ```solidity
+     * bytes32 hash = keccak256(abi.encodePacked(
+     *      \x19\x01,
+     *      CURRENT_DOMAIN_SEPARATOR,
+     *      keccak256(
+     *          abi.encode(
+     *              keccak256("PersonalSign(bytes prefixed)"),
+     *              keccak256(abi.encode("\x19Ethereum Signed Message:\n32",contents))
+     *          )
+     *      )
+     * ));
+     * ```
+     */
+    function toNestedPersonalSignHash(bytes32 separator, bytes32 contents) internal pure returns (bytes32) {
+        return
+            MessageHashUtils.toTypedDataHash(
+                separator,
+                keccak256(abi.encode(_NESTED_PERSONAL_SIGN_TYPEHASH, MessageHashUtils.toEthSignedMessageHash(contents)))
```

Comment:
I think there is an issue with this function. See how there is a hardcoded "32" in the documentation. Also see how the contents argument is `bytes32`.

`personal_sign` messages can be arbitrary length. `contents` should be `bytes`.

Note that there are two variants of [`MessageHashUtils.toEthSignedMessageHash`](https://docs.openzeppelin.com/contracts/5.x/api/utils#MessageHashUtils-toEthSignedMessageHash-bytes32-). I think the first one is there for historical purposes. It should probably be deprecated.

I think we need the tests to use the `personal_sign` RPC call to make sure this is correct.

Reply:
Yes, I actually think this is wrong.

The `personal_sign` hash is generated by the application contract. There is no need to regenerate it here, and we don't have the message contents so we can't. 

This function should look like this:

```solidity
    function toNestedPersonalSignHash(bytes32 separator, bytes32 prefixed) internal pure returns (bytes32) {
        return
            MessageHashUtils.toTypedDataHash(
                separator,
                keccak256(abi.encode(_NESTED_PERSONAL_SIGN_TYPEHASH, prefixed))
            );
    }
```

Reply:
My understanding is that it's a `bytes32 hash` too. I'd say supporting arbitrary `bytes` opens the door to more obfuscated signature attacks.

> personal_sign does support arbitrary content length, but there is not way for this content to pass through the ERC-1271 bottleneck, is there?

Also, I've seen [it used quite frequently](https://sourcegraph.com/search?q=context:global+repo:tintinweb/smart-contract-sanctuary+file:/mainnet/+/%5CbisValidSignature%5C%28bytes+/&patternType=standard&sm=0). Some contracts reference it's a "legacy" `isValidSignature` version.

Reply:
My initial comment in this thread was misguided. The issue is another one.
> The `personal_sign` hash is generated by the application contract before calling `isValidSignature`.